docs/design/adrs/2024-03-15-use-pytest-httpserver-as-a-mock-server-for-functional-testing.md

```diff
@@ -0,0 +1,97 @@
+* [pytest-httpserver](https://pypi.org/project/pytest_httpserver/)
+* [aoai-simulated-api](https://github.com/stuartleeks/aoai-simulated-api)
+
+## Decision Outcome
+
+pytest-httpserver because it meets all of our requirements
+
+### Positive Consequences
+
+* Good, because it does not introduce another programming language or runtime
+* Good, because it's well maintained
+* Good, because has built in integration with out testing library, pytest. Therefore is quick setup.
+* Good, because it does allow us to simulate all types of errors e.g. timeouts
+
+### Negative Consequences
+
+N/A
+
+## Pros and Cons of the Options
+
+### Custom Built
+
+This option requires us to build everything from scratch. It looks fairly straight forward to do based on some 
+suggestions here [https://stackoverflow.com/questions/21877387/mocking-a-http-server-in-python](https://stackoverflow.com/questions/21877387/mocking-a-http-server-in-python).
+
+* Good, because full flexibility to return anything we want, including simulating faults
+* Good, because it does not introduce another programming language or runtime
+* Bad, because a potentially larger up front cost to develop
+* Bad, because it will require maintenance
+
+### WireMock
+
+WireMock is a java lirary that can be interacted with using the [wiremock python library](https://pypi.org/project/wiremock/).
+
+* Good, because it allows us to return anything we want, including simulating faults
+* Good, because it's well maintained
+* Good, because it's quick to setup
+* Bad, because it introduces another programming runtime
+
+### pytest-httpserver
+
+pytest-httpserver is a python library that creates a mock server that can be primed with responses. It can return bad
+status codes, but it doesn't have built in funtionality to simulate faults and timeouts in the same way as WireMock can. 
+However, it does allow us to define our own custom response handler to simulate faults e.g:
+
+```python
+    def handler(_) -> werkzeug.Response:
+        time.sleep(10)
+        return werkzeug.Response({"foo": "bar"}, status=200)
+
+    httpserver.expect_request("/foobar").respond_with_handler(handler)
+```
+
+An issue has been raised [here](https://github.com/csernazs/pytest-httpserver/issues/290) to get this feature built in.
+
+* Good, because it does not introduce another programming language or runtime
+* Good, because it's well maintained
+* Good, because has built in integration with out testing library, pytest. Therefore is quick setup.
+* Good, because it does allow us to simulate all types of errors e.g. timeouts
+
+### aoai-simulated-api
+
+aoai-simulated-api is a python library build by a Microsoft employee. It aims to create a simulated API implementation
+specifically for Azure OpenAI.
+
+* Good, because it does not introduce another programming language or runtime
+* Good, because it potentially simulates the main downstream service more accurately, though we may not need this 
```

Comment:
I agree, for CWYD, we don't need the accuracy of the AOAI simulated API results. I'm happy with using pytest http server.